velox/common/compression/Compression.h

```diff
@@ -45,6 +48,201 @@ CompressionKind stringToCompressionKind(const std::string& kind);
+
+  /// Return the smallest supported compression level.
+  /// If the codec doesn't support compression level,
+  /// `kUseDefaultCompressionLevel` will be returned.
+  virtual int32_t minimumCompressionLevel() const = 0;
+
+  /// Return the largest supported compression level.
+  /// If the codec doesn't support compression level,
+  /// `kUseDefaultCompressionLevel` will be returned.
+  virtual int32_t maximumCompressionLevel() const = 0;
+
+  /// Return the default compression level.
+  /// If the codec doesn't support compression level,
+  /// `kUseDefaultCompressionLevel` will be returned.
+  virtual int32_t defaultCompressionLevel() const = 0;
+
+  /// Performs one-shot compression.
+  /// `outputLength` must first have been computed using maxCompressedLength().
+  /// The actual compressed length will be written to actualOutputLength.
+  /// Note: One-shot compression is not always compatible with streaming
+  /// decompression. Depending on the codec (e.g. LZ4), different formats may
+  /// be used.
+  virtual Expected<uint64_t> compress(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  /// One-shot decompression function.
+  /// `outputLength` must be correct and therefore be obtained in advance.
+  /// The actual decompressed length is returned.
+  /// Note: One-shot decompression is not always compatible with streaming
+  /// compression. Depending on the codec (e.g. LZ4), different formats may
+  /// be used.
+  virtual Expected<uint64_t> decompress(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength) = 0;
+
+  /// Performs one-shot compression.
+  /// This function compresses data and writes the output up to the specified
+  /// outputLength. If outputLength is too small to hold all the compressed
+  /// data, the function doesn't fail. Instead, it returns the number of bytes
+  /// actually written to the output buffer. Any remaining data that couldn't
+  /// be written in this call will be written in subsequent calls to this
+  /// function. This is useful when fixed-length compression blocks are required
+  /// by the caller.
+  /// Note: Only Gzip and Zstd codec supports this function.
+  virtual Expected<uint64_t> compressFixedLength(
+      const uint8_t* input,
+      uint64_t inputLength,
+      uint8_t* output,
+      uint64_t outputLength);
+
+  // Maximum compressed length of given input length.
+  virtual uint64_t maxCompressedLength(uint64_t inputLength) = 0;
+
+  /// Retrieves the actual uncompressed length of data using the specified
+  /// compression library.
+  /// Note: This functionality is not universally supported by all compression
+  /// libraries. If not supported, `std::nullopt` will be returned.
+  virtual std::optional<uint64_t> getUncompressedLength(
+      const uint8_t* input,
+      uint64_t inputLength) const;
```

Comment:
please describe the behavior if the data is corrupted like folly comment
```
  /**
   * Extracts the uncompressed length from the compressed data if possible.
   * If the codec doesn't store the uncompressed length, or the data is
   * corrupted it returns the given uncompressedLength.
   * If the uncompressed length is stored in the compressed data and
   * uncompressedLength is not none and they do not match a std::runtime_error
   * is thrown.
   */
  folly::Optional<uint64_t> getUncompressedLength(
      const folly::IOBuf* data,
      folly::Optional<uint64_t> uncompressedLength = folly::none) const;
```

Reply:
I feel it makes more sense to return `Expected<uint64>` for `getUncompressedLength`. If not supported, return `Status::Invalid`. If the data is corrupted, return `Status::IOError`. What do you think?

Reply:
Looks good.